### DISCUSSION TITLE
Redability Analyzer

### DISCUSSION THREAD

**Discussion by mortii** [mortii/anki-morphs#35] [Ideas]
We should define the features we want the readability analyzer to have.

**Comment by mortii** [comment]
I could probably have the ui ported within a couple of hours.

**Comment by Vilhelm-Ian** [comment]
The readability analyzer should have two features:
1. To tell you what percent of the file you know. I used this feature to select what is my next book to read
2. To generate a custom frequency list. So while you are reading a book you encounter in anki words that are likely to appear in the book

**Comment by mortii** [comment]
> The readability analyzer should have two features:
> 
>     1. To tell you what percent of the file you know. I used this feature to select what is my next book to read
> 
>     2. To generate a custom frequency list. So while you are reading a book you encounter in anki words that are likely to appear in the book

I agree. 

Have you used 'build study plan' feature? One of the things I don't like about the readability analyzer is that is seems to do _too much_, I want to simplify it and remove features if possible.

**Reply by Vilhelm-Ian** [reply]
No I haven't

**Comment by aleksejrs** [comment]
With MorphMan, I generate a separate list with an empty profile, to compare the lists to see the words and phrases that are in mature cards, but that I might not know well.

**Reply by mortii** [reply]
What kind of lists? frequency.txt, word reports, or something else?

**Reply by aleksejrs** [reply]
frequency.txt only

**Comment by aleksejrs** [comment]
> Have you used 'build study plan' feature?

I've looked at the study plan a couple of times to see which file a word came from.

**Reply by mortii** [reply]
Could you describe the study plan feature? I don't even know what it does.

**Reply by aleksejrs** [reply]
It generates `study_plan.txt`, a version of `frequency.txt` with lines like these before the entries for each file:

`'file.txt' study goal: (995/ 995) morph readability: 63.93 -> 95.00 line readabiltiy: 0.00 -> 0.00`

`'otherfile.txt' study goal: (1250/2245) morph readability: 71.39 -> 95.00 line readabiltiy: 0.00 -> 0.00`

`'thirdfile.txt' study goal: (880/3125) morph readability: 55.29 -> 95.03 line readabiltiy: 2.83 -> 77.36`

**Comment by aleksejrs** [comment]
I used to have a bunch of lists of words I would like to know better, which I used as input for grep to find the words in frequency.txt.  Because I left all of the grammatical normalization out of MorphMan, the grepping script would add prefixes and suffixes to the pattern, specific to the language-specific list.

Eventually grepping became too slow because the lists contained hundreds or thousands of words, so I switched to just looking at the top of the frequency.txts in vimdiff and ordering the corpus to get the more interesting words to the top, since the exact position of the word in the list is not so important.

**Comment by aleksejrs** [comment]
I don't know what "score" means, but "ep_freq" and "all_freq" are useful. "master_freq" would also be, if I used a master list.

Because my morphs contain spaces, the syntax used in MorphMan interferes with grepping.

**Comment by aleksejrs** [comment]
> the exact position of the word in the list is not so important.

And that is bad when I want to give some files distinct priorities. I use symlinks to increase a directory's all_freq numbers; IIRC, it works.
I wouldn't want every file to have a much higher priority than the immediately following files though.

**Comment by mortii** [comment]
Btw, any suggestions for a better name than 'Readability Analyzer'? With that name it's not immediately intuitive that it can do stuff like generate frequency.txt.

**Reply by aleksejrs** [reply]
Frequency lists

**Reply by aleksejrs** [reply]
Maybe "Word lists", if it is to include proper names and known morphs.

**Reply by Vilhelm-Ian** [reply]
Maybe MorphTools. It's not descriptive but at least it's not confusing like Readability Analyzer.

**Reply by Vilhelm-Ian** [reply]
Maybe it should not even have a name. Maybe it should just be a verb

**Comment by Vilhelm-Ian** [comment]
I think it would be better if those two features would be separate. Because I can't think a name that would describe both features

**Reply by aleksejrs** [reply]
What two features? Word lists and stats?

**Reply by Vilhelm-Ian** [reply]
yes

**Reply by mortii** [reply]
Do you mean splitting 'readability analyzer' into separate tools?

**Reply by Vilhelm-Ian** [reply]
Yes. But I am just putting it out as an idea there. Not sure how it goes. Atleast that's what you do in programming when a function  does more than one thing. My reasoning is that there is no way to describe it so it's clear what all the functionality it offers. 

We can also go the other route. Give it a name that is totally meaningless like MorphTools it's not descriptive at all. But atleast it's not confusing like readability analyzer.

**Reply by mortii** [reply]
MorphTools is a million times better name. The case for not splitting it up into several tools is that are is significant overlap  (input directory field, output directory field, etc), the only real difference is the output.

**Reply by Vilhelm-Ian** [reply]
I agree entirely with your argument.

**Reply by aleksejrs** [reply]
Note that the morphs for frequency.txt are chosen based on the file's readability, so only the top specified % of unmature morphs go into the list.

**Reply by mortii** [reply]
@aleksejrs is that a good thing? sounds unnecessarily complicated to me.

**Reply by aleksejrs** [reply]
It shortens the list by skipping some of the morphs that only occur once. Since I use two-word morphs, there are many such morphs.

**Reply by mortii** [reply]
is percent better than a 'only include morphs that occur more than x times'-option?

**Reply by aleksejrs** [reply]
For a long or low-priority file, I would skip the one-time morphs, but for a short high-priority file or a song, I would include all morphs. Having some single-instance morphs and leaving others out makes no sense to me.

**Reply by mortii** [reply]
^ I just saw the minimum frequency option, that is the equivalent to what i described right?

**Reply by aleksejrs** [reply]
No, the tooltip for "Minimum Frequency" says “corresponding to the 'Master Frequency List' in the Study Plan”.

**Reply by aleksejrs** [reply]
> for a short high-priority file or a song, I would include all morphs.

Maybe even those split across lines. I guess there are many files where a paragraph is split into lines, including subtitles. Don't forget about subtitles.

**Comment by mortii** [comment]
The more I look at it the more I agree that  readability analyzer should be split up...

I think it boils down to how many features we want. If we have 50 different features that have mostly the same layout then it makes sense to have them all at the same place, but if we only have two then they can be in their own custom made dialogs.

@aleksejrs  do you also for the most part only use the frequency.txt and readability report generators? 

I'm very tempted to scrap the other features.

**Reply by aleksejrs** [reply]
> @aleksejrs do you also for the most part only use the frequency.txt and readability report generators?

I use:
* Input Directory field
* Morphemizer selector (I never change it, but I use the same custom morphemizer for everything)..
* Target % (to shorten the list).
* frequency.txt and study plan
* the Study Plan tab. I actually look at it quite often. It shows those percentages that go into study_plan.txt (IIRC, there are more numbers than in the Readability Report tab. You should just try it). It informs me about the order of the files processed and the duplicates found in the Input Directory. It does *not* help me see what happens to the symlinks.

I don't touch the Known Morphs DB setting. I may have used the Database Manager once or twice, I don't like it.
Output Directory is probably too general.

Since I am using Anki 2.1.22 for MorphMan, I only recall that there is a checkbox for the file/directory prioritization. I had been using the analyzer for years before I noticed that the ordering is not just all_freq, but I am not sure what score is for, then.

**Comment by aleksejrs** [comment]
This may be related to my customization of MorphMan, but I feel the need delete the databases sometimes to make it regenerate some fields or something, if I removed some words from a field.

**Comment by Vilhelm-Ian** [comment]
@aleksejrs how important is to you study plan

**Reply by aleksejrs** [reply]
The columns I like in the two tables are Total Morphs, Known Morphs, Known Morphs %, To Study Morphs, (Old) Morph Readability, New Morph Readability. I don't know what "instances" are, my morphemizer does not support proper nouns, and I don't pay attention to the line stuff. Cumulative Morphs is also kinda interesting, but I don't really look at it. I prefer the Study Plan table because it is simpler, and the columns I listed are somewhat redundant together.

Readability Report:
![readabilityreport](https://github.com/mortii/anki-morphs/assets/1025861/d5c5901a-864e-4683-9b76-9c71863f604a)


Study Plan:
![studyplan](https://github.com/mortii/anki-morphs/assets/1025861/f17c4be4-6327-4aaa-bafa-ddf4ecd4bd24)

**Comment by mortii** [comment]
Ok, I have played around with the morphman readability analyzer and I have some thoughts:
- 'Readability report' is actually a cool feature, we should definitely implement that
- 'Study plan' honestly looks like it's borderline completely useless. The interesting columns can also be inferred in the readability report, the "Old *" and 'New *" columns make no sense. I don't think this is a feature worth implementing.
- The target percent option has the description: _"The target 'Readability %' for the study plan. Morphemes meeting the 'Minimum Frequency' will be added to the plan until the 'Target %' is reached."_ Since this is only related to study plan this option is now unnecessary.

This now leaves the features:
- Generate frequency list
- Generate readability report

I think these two features have enough difference in input that it's worth splitting them into two different dialogs.

**Reply by aleksejrs** [reply]
> The target percent option has the description: _"The target 'Readability %' for the study plan. Morphemes meeting the 'Minimum Frequency' will be added to the plan until the 'Target %' is reached."_ Since this is only related to study plan this option is now unnecessary.

No, it applies to frequency.txt, at least in my old MorphMan.

**Reply by mortii** [reply]
That is the copy-paste of the description. I don't think it makes sense in a frequency list generating context anyway, a minimum frequency option is sufficient imo.

**Comment by mortii** [comment]
I have  split these two features into different discussions:
- [Frequency list generator #36](https://github.com/mortii/anki-morphs/discussions/36)
- [Readability report generator #37](https://github.com/mortii/anki-morphs/discussions/37)
